Initial implementation of asset bundle management

randomPoison opened the pull request:
> NOTE: This PR is not yet fully ready to be merged (specifically it is missing some polish work like documentation), but the core design and implementation is in place and ready for scrutiny.

This PR attempts to define a minimal, extensible set of core functionality for managing asset bundles in the form of a new asset-bundle-builder Unity package. This package does two key things:

* It defines a set of conventions for how asset bundles are defined, named, hosted, and identified at runtime.
* It provides core utilities for implementing these conventions in a Unity project.

[`README.md`](https://github.com/kongregate/asset-bundle-builder/pull/1/files#diff-04c6e90faac2675aa89e2176d2eec7d8) contains the usage documentation and the core specification of the system, and is a good place to start when reviewing. Please also review the code API in order to verify that it meets the requirements laid out in the specification.

randomPoison committed:
Add current version of the package

randomPoison committed:
Document AssetBundleDescription

randomPoison committed:
Change package name to asset-bundle-builder

randomPoison committed:
Make hash and default constructor

randomPoison committed:
Update README.md

randomPoison committed:
Cleanup AssetBundleTarget enum

* Remove incorrect WSAPlayer variant.
* Remove explicit values from variants.
* Remove inaccurate paragraph in documentation.

randomPoison committed:
Update documentation for AssetBundleTarget

randomPoison committed:
Add Json.NET dependency and update setup instructions

randomPoison committed:
Document AssetBundleBuilder

randomPoison committed:
More updates/tweaks to docs and comments

randomPoison committed:
Add MIT license

randomPoison committed:
Ignore computed properties when serializing JSON

randomPoison committed:
Add dependencies to bundle description list

randomPoison committed:
Update README.md

randomPoison committed:
Rename CachedAssetBundle to CachedAssetBundleForCurrentTarget

roberson3d commented:
Do you know how well this will work with the new Addressables system? From what I can tell Addressables are supports similar things: figuring out local vs distributed, what's new vs unchanged, handle building the bundles, etc. It would be good to find out what overlap, if any, there is as I would image moving towards using the Addressables systems in the future.

randomPoison commented:
I did some research into Addressables. From what I can tell, this system wouldn't be directly compatible with Addressable and it does try to emulate a lot of the functionality that Addressables provides. In theory, it would be better to just use Addressables, but in practice this package still has use for games that are already using asset bundles. Based on internal discussion, it seems like this system will still be useful for games that are currently in development that want to continue to use the asset bundle development workflow.

I'll add a note about compatibility with Addressables to the README for future reference.

randomPoison committed:
Add notes about addressables

roberson3d commented on Runtime/AssetBundleTarget.cs:
```diff
@@ -0,0 +1,30 @@
+    /// </remarks>
+    public enum AssetBundleTarget
+    {
+        StandaloneOSX,
```
What's the purpose of having the "Standalone" prefix for some targets if it doesn't differentiate between the editor and standalone players?

randomPoison replied:
It might not be necessary 🤔 For OSX and Linux it would probably be fine to drop the "Standalone" prefix, but for Windows there's both "Standalone Windows" and "Windows Store Application" which use different bundles.

Also, while the build system doesn't differentiate between editor and standalone builds, it is *using* the standalone asset bundles in the editor. So while it may not be strictly necessary to include the "Standalone" prefix, it more accurately indicates which Unity platform is being used.

Not sure if those are super good reasons for keeping the "Standalone" prefix, though 😅 Honestly, I'm not even sure the `AssetBundleTarget` enum makes sense or is necessary. It seems like `RuntimeTarget` *mostly* covers what we need, except that it has different variants for the editors. Would it maybe be better to just use `RuntimePlatform` and provide a utility for normalizing which variant is used on each platform?

randomPoison replied:
I removed `AssetBundleTarget` entirely in favor of using `RuntimePlatform`. Based on what [a Unity developer mentioned on the forums](https://forum.unity.com/threads/do-macos-and-windows-need-different-asset-bundles.670510/#post-4881242), it seems like `RuntimePlatform`'s variants line up with the distinct platforms that need separate asset bundles. In order to correctly handle editor support, the helper methods/properties on `AssetBundleDescription` automatically normalize the provided platform.

Does this seem good to you?

roberson3d replied:
👍 makes sense to me

roberson3d commented on Runtime/AssetBundleDescription.cs:
```diff
@@ -0,0 +1,277 @@
+        /// platforms will be considered supported, such that the bundle cannot be
+        /// loaded on any platforms not specified.
+        /// </param>
+        [JsonConstructor]
```
JSON.net isn't our standard json parsing tool. For games that aren't using it, would the expectation be to use it just for this feature or to manually construct the needed data using the standard parsing tool of the game?

randomPoison replied:
This is a tough one. In theory, I'd like to not require Json.NET as part of the main package, that way projects that don't use Json.NET don't have to pull in the dependency. In practice though, there's no other way to provide out-of-the-box JSON serialization support for this data. The expectation is that if you want to use custom serialization (e.g. for serializing to a different format) you'll have to write that logic yourself, which would include using a different JSON serialization process than Json.NET.

The main reason why I included Json.NET directly, rather than letting it be optional, was so that I could use the Json.NET attributes to customize serialization. I think I might be able to provide that same functionality via a separate package, that way asset-bundle-builder won't have to depend on Json.NET directly. Let me try that out and see if I can get it working 🙂

randomPoison replied:
I've made Json.NET an optional dependency. It's disabled by default and can be enabled by adding a `JSON_NET` definition to the project. For games that aren't using JSON or aren't using Json.NET, the raw information in `AssetBundleDescription` can be used directly to serialize out to the desired format.

randomPoison committed:
Make Json.NET dependency optional

* Remove JSON attributes from AssetBundleDescription.
* Add new AssetBundleDescriptionConverter class to handle serialization.
* Remove Hash128Converter since it's no longer needed.
* Remove jillejr.newtonsoft.json-for-unity package dependency.
* Move Json.NET functionality into new assembly and setup JSON_NET define constraint.
* Make AssetBundleBuilder a struct and make sure field access is immutable.
* Update the README with instructions for how to enable Json.NET support.

randomPoison committed:
Remove AssetBundleTarget enum

Replace AssetBundleTarget with RuntimePlatform, and add in necessary logic to normalize the platform to handle the editor correctly.

randomPoison committed:
Remove instructions for setting up NPM registry

randomPoison committed:
Update notes on platform handling

randomPoison committed:
Rename FileNameForCurrentTarget and HashForCurrentTarget

* FileNameForCurrentTarget -> FileName.
* HashForCurrentTarget -> Hash.

randomPoison committed:
Minor wording update

ajarms reviewed (approved):
This setup looks good
